examples/pytorch/gat_pyg.py

```diff
@@ -0,0 +1,177 @@
+        self.num_classes = num_classes
+        self.out_dim = num_classes
+        self.xent = nn.CrossEntropyLoss()
+        self.conv1 = GATConv(
+            in_channels=in_dim,
+            out_channels=hidden_dim,
+            heads=head_num[0],
+            dropout=0.6,
+        )
+        layer0_output_dim = head_num[0] * hidden_dim
+        self.conv2 = GATConv(
+            in_channels=layer0_output_dim,
+            out_channels=self.out_dim,
+            heads=1,
+            dropout=0.6,
+            concat=False,
+        )
+
+    def forward(self, inputs):
+        """Calculate loss, make predictions and fetch labels."""
+        nodes, feat, edge_index, mask, label = inputs
+        nodes = torch.squeeze(nodes.to(torch.int32))  # [N]
+        feat = torch.squeeze(feat.to(torch.float32))  # [N, F]
+        edge_index = torch.squeeze(edge_index.to(torch.int32))  # [2, X]
+        mask = torch.squeeze(mask.to(torch.bool))  # [N]
+        labels = torch.squeeze(label.to(torch.int32))  # [N]
+
+        x = feat
+        x = F.dropout(x, p=0.6, training=self.training)
+        x = F.elu(self.conv1(x, edge_index))
+        x = F.dropout(x, p=0.6, training=self.training)
+        scores = self.conv2(x, edge_index)
+        labels = labels.type(torch.int64)
+        labels = labels[mask]  # [batch_size]
+        scores = scores[mask]  # [batch_size]
+        loss = self.xent(scores, labels)
+
+        return loss, scores, labels
+
+
+def train_func(config: dict):
+    p = GATQueryParameter(
+        neighbor_edge_types=np.array([0], np.int32),
+        feature_idx=config["feature_idx"],
+        feature_dim=config["feature_dim"],
+        label_idx=config["label_idx"],
+        label_dim=config["label_dim"],
+    )
+    model = GAT(
+        in_dim=config["feature_dim"],
+        head_num=[8, 1],
+        hidden_dim=8,
+        num_classes=config["num_classes"],
+        q_param=p,
+    )
+
+    optimizer = torch.optim.Adam(
+        filter(lambda p: p.requires_grad, model.parameters()),
+        lr=0.005,  # * session.get_world_size(),
+        weight_decay=0.0005,
+    )
+
+    g = Cora()
+    train_np = np.loadtxt(f"{g.data_dir()}/train.nodes", dtype=np.int64)
+    eval_batch = np.loadtxt(f"{g.data_dir()}/test.nodes", dtype=np.int64)
+
+    model.train()
+    train_tensor = model.query(g, train_np)
+    torch_tensor = [torch.from_numpy(a) for a in train_tensor]
+    for _ in range(config["num_epochs"]):
+        loss, _, _ = model(torch_tensor)
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+
+    model.eval()
+
+    eval_tensor = model.query(g, eval_batch)
+    _, scores, label = model([torch.from_numpy(a) for a in eval_tensor])
+    accuracy = torch.tensor(
+        accuracy_score(
+            y_true=label.cpu(), y_pred=scores.argmax(dim=1).detach().cpu().numpy()
+        )
+    ).item()
+    assert accuracy == 0.828
+
+
+if __name__ == "__main__":
+    train_func(
```

Comment:
why is train_func not in main?

Reply:
why does it have to be main? Please provide a reference to a style guide.